cli/src/semgrep/engine.py

```diff
@@ -33,52 +33,54 @@ class EngineType(Enum):
     ) -> "EngineType":
         """Select which Semgrep engine type to use if none is explicitly requested.
 
         Considers settings from Semgrep Cloud Platform and version control state.
         """
-        # Change default to pro-engine intrafile if secrets was requested.
-        # Secrets is built into pro-engine, but any pro-setting should work.
-        if (
-            not (scan_handler and scan_handler.deepsemgrep)
-            and requested_engine is None
-            and run_secrets
-        ):
-            requested_engine = cls.PRO_LANG
-        elif run_secrets and requested_engine is cls.OSS:
-            # Should be impossible if the CLI gates impossible arguement combinations.
-            raise SemgrepError("Semgrep Secrets is not part of the open source engine")
+        interfile_is_requested_via_app = scan_handler and scan_handler.deepsemgrep
 
-        if git_meta and scan_handler:
-            if scan_handler.deepsemgrep and requested_engine is None:
+        requested_engine = engine_flag
+        if engine_flag is None:
+            if interfile_is_requested_via_app:
                 requested_engine = cls.PRO_INTERFILE
-
-            if (
-                requested_engine == cls.PRO_INTERFILE
-                and not git_meta.is_full_scan
-                and not enable_pro_diff_scan
-            ):
+            elif run_secrets:
                 requested_engine = cls.PRO_INTRAFILE
 
-        # Using PRO_LANG engine since PRO_INTERFILE/PRO_INTRAFILE defaults to -j 1
-        # note if using OSS, then will keep using OSS
+        if run_secrets and engine_flag is cls.OSS:
+            # Should be impossible if the CLI gates impossible argument combinations.
+            # TODO Can we delete this check?
+            raise SemgrepError("Semgrep Secrets is not part of the open source engine")
+
+        diff_scan = git_meta and not git_meta.is_full_scan
         if (
-            requested_engine in {cls.PRO_INTERFILE, cls.PRO_INTRAFILE}
-            and supply_chain_only
+            diff_scan
+            and not interfile_diff_scan_enabled
+            and requested_engine is cls.PRO_INTERFILE
         ):
-            logger.info(
-                "Running only supply chain rules so running without extra interfile analysis"
-            )
-            return cls.PRO_LANG
+            requested_engine = cls.PRO_INTRAFILE
+
+        # Hack: Turn off the PRO_INTERFILE engine when only supply chain is requested
+        # This is necessary because PRO_INTERFILE defaults to `-j 1`
+        if supply_chain_only and requested_engine is cls.PRO_INTERFILE:
+            requested_engine = cls.PRO_INTRAFILE
 
-        return requested_engine or cls.OSS
+        # TODO: should we fail here if logged_in is false and requested_engine is not cls.OSS?
+
+        # `logged_in and ci_scan_handler` is redundant because `ci_scan_handler` requires
+        # being logged in, but let's check this defensively
```

Comment:
Why do we need to be defensive about `scan_handler`?

Reply:
Fleshed out the comment. Basically I want to separate the concerns of "is this a ci scan?" and "is this a logged in scan?". In particular, I wouldn't want someone to change how `scan_handler` works and then accidentally allow logged out pro scans. This is not possible right now for various reasons but I want to be explicit since it's kind of important.